### ISSUE TITLE
HTTP Compression - Request Compression

### ISSUE THREAD

**Issue by EatonZ** [dotnet/runtime#46944]
## Background and Motivation

There is currently no support for HTTP **request** compression. Response decompression support has been present and maintained for many years now, but request compression has been oddly absent.

## Proposed API

Add a `CompressedContent` class. An example of an existing implementation is [here](https://github.com/WebApiContrib/WebAPIContrib/blob/master/src/WebApiContrib/Content/CompressedContent.cs). It needs a refresh for 2021 (Brotli support, etc) but otherwise showcases support is possible, and easy to implement.

The CompressedContent name also seems the best to use - it goes along with [DecompressedContent](https://github.com/dotnet/runtime/blob/0c7dde5de145940dc19bda59a32554ffcafc8c73/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/DecompressionHandler.cs#L96).

## Usage Examples

A snippet of code from my own project, which uses a modified version of the WebAPIContrib `CompressedContent`. `content` is one of `ByteArrayContent`, `FormUrlEncodedContent`, `StreamContent`, `StringContent`, or any other class inheriting `HttpContent`.

```
var request = new HttpRequestMessage(RequestMethod, URI);
request.Content = RequestDetails.RequestCompression switch
{
    DecompressionMethods.GZip => new CompressedContent(content, "gzip"),
    DecompressionMethods.Deflate => new CompressedContent(content, "deflate"),
    DecompressionMethods.Brotli => new CompressedContent(content, "br"),
    DecompressionMethods.All => throw new Exception("EasyHttp.HttpExecuteAsync: Only 1 request compression method can be specified."),
    _ => content
};
var response = await httpClient.SendAsync(request);
```

## Alternative Designs

N/A at this time.

## Risks

None seen since this would be opt-in.

**Comment by karelz** [comment]
I assume this is achievable today by setting the headers and manually compressing the content. It probably also allows better flexibility in tweaking compression algorithm parameters (size vs. time tradeoff).
I think it will boil down to how often is the scenario needed.

We have similar new compression API on WebSocket: https://github.com/dotnet/runtime/issues/31088#issuecomment-754854011

**Comment by EatonZ** [comment]
@karelz Yes, it can be done today if you copy that example `CompressedContent` class into your project. I just feel it should be built-in so you don't need to do that.

Compression params would be easy - just add `CompressionLevel` to the constructor to let the user specify their preference for speed or size.

I am working on a modernized version of `CompressedContent` for my own project. If the .NET team would approve adding it, I'd be happy to put in a PR, but I'd need to know _where_ this class would best fit.

**Comment by karelz** [comment]
@EatonZ we first need agreement on Networking team this is something valuable for enough customers (in our triage) and then we will need official API approval (see [API review process](https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md), before we would be ready for a PR.
Expect it will take some time - also to let more folks to comment and not to rush it without considering alternatives / need.

**Comment by EatonZ** [comment]
Here's a complete example of how it could look:

```
public sealed class CompressedContent : HttpContent
{
    private readonly HttpContent originalContent;
    private readonly string contentEncoding;
    private readonly CompressionLevel compressionLevel;

    public CompressedContent(HttpContent originalContent, string contentEncoding, CompressionLevel compressionLevel)
    {
        this.originalContent = OriginalContent;
        this.contentEncoding = ContentEncoding;
        this.compressionLevel = CompressionLevel;

        foreach (var (key, value) in originalContent.Headers) Headers.TryAddWithoutValidation(key, value);
        Headers.ContentEncoding.Add(contentEncoding);
    }

    protected override async Task SerializeToStreamAsync(Stream stream, TransportContext context)
    {
        Stream compressedStream = contentEncoding switch
        {
            "br" => new BrotliStream(stream, compressionLevel, true),
            "gzip" => new GZipStream(stream, compressionLevel, true),
            "deflate" => new DeflateStream(stream, compressionLevel, true)
        };
        await using (compressedStream.ConfigureAwait(false))
        {
            await originalContent.CopyToAsync(compressedStream).ConfigureAwait(false);
        }
    }

    protected override bool TryComputeLength(out long length)
    {
        length = 0;
        return false;
    }
}
```

**Comment by karelz** [comment]
Triage: May be worthwhile, but only 1 customer asking for it + there is simple workaround. Let's wait if more people upvote the top post before adding yet another API.

**Comment by samsosa** [comment]
Most RESTful API implementations support implicit `Content-Encoding: gzip`. 
We use it for all such kind API services, as it significantly reduces the amount 
of JSON-Payload transferred. (... by up to 90 %)

Currently, we compress the payload before sending it using GZipStream and 
set the appropriate request header.
We would not reject a clean, reasonable API within .NET. 😉

**Comment by EatonZ** [comment]
@karelz Any interest in getting this into .NET 6?

**Comment by scalablecory** [comment]
@EatonZ we won't be able to get this in for .NET 6. we still need agreement to add the feature -- if you can help drive this conversation, that'd be helpful! then, we need API review before getting a PR.

**Comment by EatonZ** [comment]
[Request decompression support was added in the recently released .NET 7](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/middleware/request-decompression?view=aspnetcore-7.0). I think request compression should definitely be added now to compliment that feature.🙂

**Comment by bartonjs** [comment]
[Video](https://www.youtube.com/watch?v=oguJUzjKP1c&t=1h1m26s)

* Rather than an enum-based dispatcher, let's just expose each type as a distinct HttpContent type.
* Let's leave Deflate out since it is unlikely to be needed.

```csharp
namespace System.Net.Http
{
    public sealed class ZstandardCompressedContent : System.Net.Http.HttpContent
    {
        public ZstandardCompressedContent(HttpContent content) { }
        public ZstandardCompressedContent(HttpContent content, ZstandardCompressionOptions compressionOptions) { }
    }

    public sealed class BrotliCompressedContent : System.Net.Http.HttpContent
    {
        public BrotliCompressedContent(HttpContent content) { }
        public BrotliCompressedContent(HttpContent content, BrotliCompressionOptions compressionOptions) { }
    }

    public sealed class GZipCompressedContent : System.Net.Http.HttpContent
    {
        public GZipCompressedContent(HttpContent content) { }
        public GZipCompressedContent(HttpContent content, ZLibCompressionOptions compressionOptions) { }
    }
}
```

**Comment by rzikm** [comment]
The *CompressionOptions types are a bit low level and it may not be obvious for users how to configure them properly. The original proposal with `CompressionLevel compressionLevel` (the enum with `Fastest`, `Optimal`, `SmallestSize` members, not to be confused with integer parameter of the same name we have for zlib-based compression APIs) provided a more high-level, easier to use turning knob.

Do we care? Maybe add it as a default param to the first ctor like `public GZipCompressedContent(HttpContent content, CompressionLevel compressionLevel = CompressionLevel.Optimal) { }`?

**Comment by bartonjs** [comment]
I was under the impression that the CompressionLevel enum mapped to those options types cleanly, but I see that isn't true for ZLibCompressionOptions, at least.

Personally, I'd be fine with it either as a defaulted parameter (which has to be done before RC1 snap... assuming you add these types before the RC1 snap) or as an overload (which can be added whenever).

```csharp
    public sealed class GZipCompressedContent : System.Net.Http.HttpContent
    {
        public GZipCompressedContent(HttpContent content, CompressionLevel compressionLevel = CompressionLevel.Optimal) { }
        public GZipCompressedContent(HttpContent content, ZLibCompressionOptions compressionOptions) { }
    }
```

or

```csharp
    public sealed class GZipCompressedContent : System.Net.Http.HttpContent
    {
        public GZipCompressedContent(HttpContent content) { }
        public GZipCompressedContent(HttpContent content, CompressionLevel compressionLevel) { }
        public GZipCompressedContent(HttpContent content, ZLibCompressionOptions compressionOptions) { }
    }
```

Or even

```csharp
    public partial class ZLibCompressionOptions
    {
        /// <remarks>
        ///   Translates the .NET platform concept of a compression level into
        ///   zlib's concept of compression level.  Not to be confused with
        ///   the <see cref="CompressionLevel" /> property.
        /// </remarks>
        public ZLibCompressionOptions(CompressionLevel compressionLevel);
    }
```

If y'all are interested in either of the first two, just send it in an email, it's a small enough delta that I think we can handle it without a live meeting.

**Comment by ManickaP** [comment]
I like the first option the most `public GZipCompressedContent(HttpContent content, CompressionLevel compressionLevel = CompressionLevel.Optimal) { }`.

**Comment by iremyux** [comment]
@bartonjs , as we discussed it offline, we decided on this design for each class: 

```c#
public sealed class GZipCompressedContent : System.Net.Http.HttpContent
{
    public GZipCompressedContent(HttpContent content, CompressionLevel compressionLevel = CompressionLevel.Optimal) { }
    public GZipCompressedContent(HttpContent content, ZLibCompressionOptions compressionOptions) { }
}
```

The implementation is already merged with PR #130082
